x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/ingest/GeoTileProcessor.java

```diff
@@ -0,0 +1,421 @@
+                }
+            }
+        } catch (Exception e) {
+            throw new IllegalArgumentException("invalid tile definition", e);
+        }
+
+        return ingestDocument;
+    }
+
+    /**
+     * This class handles the different tile types, rectangular or hexagonal.
+     * Each of geohash, geotile and geohex have different ways of describing polygons, and other attributes.
+     */
+    abstract static class TileHandler {
+        Geometry makePolygonFrom(Rectangle r) {
+            double[] x = new double[] { r.getMinX(), r.getMaxX(), r.getMaxX(), r.getMinX(), r.getMinX() };
+            double[] y = new double[] { r.getMinY(), r.getMinY(), r.getMaxY(), r.getMaxY(), r.getMinY() };
+            Polygon polygon = new Polygon(new LinearRing(x, y));
+            if (GeometryNormalizer.needsNormalize(Orientation.CCW, polygon) == false) {
+                return polygon;
+            }
+            return GeometryNormalizer.apply(Orientation.CCW, polygon);
+        }
+
+        abstract Geometry makePolygon();
+
+        abstract String[] makeChildren();
+
+        // Most tilers have no intersecting non-children
+        String[] makeNonChildren() {
+            return new String[] {};
+        }
+
+        abstract int getDepth();
+
+        abstract String getParent();
+    }
+
+    static class GeotileHandler extends TileHandler {
+        private final String address;
+        private final int zoom;
+        private final int x;
+        private final int y;
+
+        GeotileHandler(String address) {
+            this.address = address;
+            final int[] hashAsInts = GeoTileUtils.parseHash(address);
+            this.zoom = hashAsInts[0];
+            this.x = hashAsInts[1];
+            this.y = hashAsInts[2];
+        }
+
+        @Override
+        Geometry makePolygon() {
+            return makePolygonFrom(GeoTileUtils.toBoundingBox(address));
+        }
+
+        @Override
+        String[] makeChildren() {
+            int z = zoom + 1;
+            if (z > GeoTileUtils.MAX_ZOOM) {
+                return new String[] {};
+            }
+            int xo = x * 2;
+            int yo = y * 2;
+            return new String[] { hash(z, xo, yo), hash(z, xo + 1, yo), hash(z, xo, yo + 1), hash(z, xo + 1, yo + 1) };
+        }
+
+        private static String hash(int zoom, int x, int y) {
+            return zoom + "/" + x + "/" + y;
+        }
+
+        @Override
+        String getParent() {
+            return zoom == 0 ? "" : hash(zoom - 1, x / 2, y / 2);
+        }
+
+        @Override
+        int getDepth() {
+            return zoom;
+        }
+    }
+
+    static class GeohashHandler extends TileHandler {
+        private final String hash;
+
+        GeohashHandler(String hash) {
+            this.hash = hash;
+        }
+
+        @Override
+        Geometry makePolygon() {
+            return makePolygonFrom(Geohash.toBoundingBox(hash));
```

Comment:
As with GeoTile, it should be a Rectangle geometry

Reply:
Done